src/ignorer.rs

```diff
@@ -0,0 +1,71 @@
+use {
+    anyhow::{Context, Result},
+    git_repository::{
+        self as git,
+        prelude::FindExt,
+        Repository,
+    },
+    std::path::{Path, PathBuf},
+};
+
+/// An object able to tell whether a file is excluded
+/// by gitignore rules
+pub struct Ignorer {
+    repo: Repository,
+}
+
+impl Ignorer {
+
+    /// root_path is assumed to exist and be a directory
+    pub(crate) fn new(root_path: &Path) -> Result<Self> {
+        let repo = git::discover(root_path)?;
+        Ok(Self { repo })
+    }
+
+    /// Tell whether the given path is excluded according to
+    /// either the global gitignore rules or the ones of the repository
+    pub fn excludes(&mut self, file_path: &Path) -> Result<bool> {
+        // TODO this code does work but is wasteful (creating a cache
+        // for each file, for example). I hope this should get better
+        // when gitoxide matures
+
+        let worktree = self.repo.worktree().context("a worktree should exist")?;
+        let index = worktree.index()?;
+
+        // there doesn't seem to be any public API for looking at "excludes" without caching
+        // so we create a cache
```

Comment:
Cache is probably a misnomer here as it doesn't cache anything - it merely builds state to be able to perform these lookups and be fast with non-random input ordered according to the git index.
I think I ran out of words as it contains an ignore `stack` internally, and this is adding even more information to bring everything together.